From 0c4133a38b3fa92afc56faf7e0f05256a77d6b20 Mon Sep 17 00:00:00 2001
From: Richard Gong <richard.gong@intel.com>
Date: Tue, 15 May 2018 13:13:52 -0500
Subject: [PATCH 01/18] FogBugz #559704-3: update RSU driver to align with
 changes at svc

commit  f52856f3231dd2f7928b9fe4220fe768982b0014 from
https://github.com/altera-opensource/linux-socfpga.git

Intel-rsu driver call service layer to free service layer resources when
status or update request is completed, or there is an error in status or
update request process.

Signed-off-by: Richard Gong <richard.gong@intel.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/misc/intel-rsu.c | 26 ++++++++++++++++++--------
 1 file changed, 18 insertions(+), 8 deletions(-)

diff --git a/drivers/misc/intel-rsu.c b/drivers/misc/intel-rsu.c
index 4dc834b68134..3070d931d54b 100644
--- a/drivers/misc/intel-rsu.c
+++ b/drivers/misc/intel-rsu.c
@@ -96,7 +96,7 @@ static int get_status(struct intel_rsu_priv *priv)
 	msg.command = COMMAND_RSU_STATUS;
 	ret = intel_svc_send(priv->chan, &msg);
 	if (ret < 0)
-		return ret;
+		goto status_done;
 
 	timeout = msecs_to_jiffies(SVC_RSU_REQUEST_TIMEOUT_MS);
 	ret =
@@ -105,15 +105,20 @@ static int get_status(struct intel_rsu_priv *priv)
 	if (!ret) {
 		dev_err(priv->client.dev,
 			"timeout waiting for COMMAND_RSU_STATUS\n");
-		return -ETIMEDOUT;
+		ret = -ETIMEDOUT;
+		goto status_done;
 	}
 	if (ret < 0) {
 		dev_err(priv->client.dev,
 			"error (%d) waiting for COMMAND_RSU_STATUS\n", ret);
-		return ret;
+		goto status_done;
 	}
 
-	return 0;
+	ret = 0;
+
+status_done:
+	intel_svc_done(priv->chan);
+	return ret;
 }
 
 /* current_image_show() - DEVICE_ATTR callback to show current_image status */
@@ -229,7 +234,7 @@ static int send_update(struct intel_rsu_priv *priv,
 
 	ret = intel_svc_send(priv->chan, &msg);
 	if (ret < 0)
-		return ret;
+		goto update_done;
 
 	timeout = msecs_to_jiffies(SVC_RSU_REQUEST_TIMEOUT_MS);
 	ret = wait_for_completion_interruptible_timeout(&priv->svc_completion,
@@ -237,15 +242,20 @@ static int send_update(struct intel_rsu_priv *priv,
 	if (!ret) {
 		dev_err(priv->client.dev,
 			"timeout waiting for COMMAND_RSU_UPDATE\n");
-		return -ETIMEDOUT;
+		ret = -ETIMEDOUT;
+		goto update_done;
 	}
 	if (ret < 0) {
 		dev_err(priv->client.dev,
 			"error (%d) waiting for COMMAND_RSU_UPDATE\n", ret);
-		return ret;
+		goto update_done;
 	}
 
-	return 0;
+	ret = 0;
+
+update_done:
+	intel_svc_done(priv->chan);
+	return ret;
 }
 
 /* reboot_image_store() - DEVICE_ATTR callback to store reboot_image request */
-- 
2.17.1

